compiler/rustc_middle/src/thir.rs

```diff
@@ -811,28 +811,44 @@ pub enum PatKind<'tcx> {
     ///   exhaustiveness to cover exactly its own value, similar to `&str`, but these values are
     ///   much simpler.
     /// * `String`, if `string_deref_patterns` is enabled.
+    ///
+    /// # Caution
+    ///
+    /// Note that in general, "constants" don't necessarily get lowered to this!
+    /// For example, a constant of an enum type would typically be lowered to
+    /// `PatKind::Variant` instead.
     Constant {
         value: mir::Const<'tcx>,
     },
 
-    /// Pattern obtained by converting a constant (inline or named) to its pattern
-    /// representation using `const_to_pat`.
+    /// Marker node that wraps the result of converting some kind of "constant"
```

Comment:
How about something like: "A constant in a pattern gets expanded into a `thir::Pat` that exactly matches its value. For non-trivial constants (i.e. those that aren't literals handled by PatKind::Constant), we wrap it in an `ExpandedConstant` node. Unsafeck and some diagnostics rely on the presence of this node." and then the note about range patterns

Reply:
Since apparently we're not adding this node for all constants, we could add "byte string literals are another exception: we don't add an `ExpandedConstant` node today.